asserts/database.go

```diff
@@ -176,3 +192,94 @@ func (db *Database) Check(assert Assertion) error {
 	}
 	return fmt.Errorf("failed signature verification: %v", lastErr)
 }
+
+func (db *Database) readAssertion(assertType AssertionType, primaryPath string) (Assertion, error) {
+	encoded, err := db.readEntry(assertionsRoot, string(assertType), primaryPath)
+	if os.IsNotExist(err) {
+		return nil, ErrNotFound
+	}
+	if err != nil {
+		return nil, fmt.Errorf("broken assertion storage, failed to read assertion: %v", err)
+	}
+	assert, err := Decode(encoded)
+	if err != nil {
+		return nil, fmt.Errorf("broken assertion storage, failed to decode assertion: %v", err)
+	}
+	return assert, nil
+}
+
+func (db *Database) readCurrentRevision(assertType AssertionType, primaryPath string) (int, error) {
+	assert, err := db.readAssertion(assertType, primaryPath)
+	if err == ErrNotFound {
+		return -1, nil
+	}
+	if err != nil {
+		return -1, err
+	}
+	return assert.Revision(), nil
+}
+
+// Add persists the assertions after ensuring it is properly signed and consistent with all the stored knowledge.
+// Returns ErrNotSuperseding if trying to add a previous revision of the assertion.
+func (db *Database) Add(assert Assertion) error {
+	reg, err := checkAssertType(assert.Type())
+	if err != nil {
+		return err
+	}
+	err = db.Check(assert)
+	if err != nil {
+		return err
+	}
+	primaryKey := make([]string, len(reg.primaryKey))
+	for i, k := range reg.primaryKey {
+		keyVal := assert.Header(k)
+		if keyVal == "" {
+			return fmt.Errorf("missing primary key header: %v", k)
+		}
+		// safety against '/' etc
+		primaryKey[i] = url.QueryEscape(keyVal)
+	}
+	primaryPath := filepath.Join(primaryKey...)
+	curRev, err := db.readCurrentRevision(assert.Type(), primaryPath)
+	if err != nil {
+		return err
+	}
+	if curRev >= assert.Revision() {
+		return ErrNotSuperseding
```

Comment:
Can we have something explicit here until we need the exported variable:

    return fmt.Errorf("assertion added must be more recent than current one (adding %d, currently %d)", ...)